In [1]:
import numpy as np

In [88]:
y = [12, 34, 29, 38, 34, 51, 29, 34, 47, 34, 55, 94, 68, 81]
y1 = [50,40,30,20,10,5,4,3,2,1,0]
y2 = [1,3,5,8,10,15,19,25,50,100]

In [89]:
type(y)

list

In [90]:
x = []
x.append(range(len(y)))

In [91]:
x

[range(0, 14)]

In [92]:
x.append([1 for ele in range(len(y))])

In [93]:
x

[range(0, 14), [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]

In [94]:
y = np.matrix(y).T

In [95]:
y

matrix([[12],
        [34],
        [29],
        [38],
        [34],
        [51],
        [29],
        [34],
        [47],
        [34],
        [55],
        [94],
        [68],
        [81]])

In [96]:
x = np.matrix(x).T

In [97]:
x

matrix([[ 0,  1],
        [ 1,  1],
        [ 2,  1],
        [ 3,  1],
        [ 4,  1],
        [ 5,  1],
        [ 6,  1],
        [ 7,  1],
        [ 8,  1],
        [ 9,  1],
        [10,  1],
        [11,  1],
        [12,  1],
        [13,  1]])

In [98]:
betas = ((x.T*x).I*x.T*y)

In [99]:
betas

matrix([[ 4.32527473],
        [17.6       ]])

matrix([[  4.32527473],  #coefficient on the time variable
        [ 17.6       ]]) #coefficient on the intercept 
        # Since the coefficient on the trend variable is positive, observations in your variable are increasing over time

In [29]:
x = range(1,len(y)+1)

In [20]:
x

range(1, 15)

In [21]:
def var(X):
    S = 0.0
    SS = 0.0
    for x in X:
        S += x
        SS += x*x
    xbar = S/float(len(X))
    return (SS - len(X) * xbar * xbar) / (len(X) -1.0)

In [22]:
def cov(X,Y):
    n = len(X)
    xbar = sum(X) / n
    ybar = sum(Y) / n
    return sum([(x-xbar)*(y-ybar) for x,y in zip(X,Y)])/(n-1)

In [23]:
def beta(x,y):
    return cov(x,y)/var(x)

In [31]:
print(beta(x,y))

-4.818181818181818


## OLS

In [100]:
def betacoeff(y):
    x = range(1,len(y)+1)
    def var(X):
        S = 0.0
        SS = 0.0
        for x in X:
            S += x
            SS += x*x
        xbar = S/float(len(X))
        return (SS - len(X) * xbar * xbar) / (len(X) -1.0)
    def cov(X,Y):
        n = len(X)
        xbar = sum(X) / n
        ybar = sum(Y) / n
        return sum([(x-xbar)*(y-ybar) for x,y in zip(X,Y)])/(n-1)
    return cov(x,y)/var(x)

In [112]:
y= [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]  #constant line
y1 = [50,40,30,20,10,5,4,3,2,1,0] #decreasing line
y2 = [1,3,5,8,10,15,19,25,50,100] #increasing line
y3 = [12, 34, 29, 38, 34, 51, 29, 34, 47, 34, 55, 94, 68, 81] #shuffled line
y4 = [1,2,3,4,5,6,7,8,9]
y5 = [9,8,7,6,5,4,3,2,1]
y6 = [-10000,500,40,30,20,10,5,4,3,2,1,0,10,40,400,4000]
y7 = [100,50,40,30,20,10,5,6,9,14,15,20]

In [113]:
[betacoeff(y),
betacoeff(y1),
betacoeff(y2),
betacoeff(y3),
betacoeff(y4),
betacoeff(y5),
betacoeff(y6),
betacoeff(y7)]

[0.0,
 -4.818181818181818,
 8.23030303030303,
 4.325274725274725,
 1.0,
 -1.0,
 306.3602941176471,
 -5.346153846153846]